### PR TITLE
Dev/nom7 v0.1

### PR DESCRIPTION
**Pull request by chifflier**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4165

Describe changes:
I started working on the upgrade to nom 7, trying to find a way to do this incrementally, and avoiding a monster-commit changing all the tree files at once. This also means that there can be more than one person working on the upgrade, since parser and dependencies are isolated in their namespaces.

- Thanks to the [`cargo` package renaming feature](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml), the package is added as `nom7`. This allows to use it as `use nom7::<items to import>;`, and more importantly it allows both versions to coexist without side-effect in the dependency tree
- nom 7 has finally removed macros, so this PR rewrites the parsers to use functions
- I ran `rustfmt` on the modified files (@victorjulien told me it was planned)
- The changes are on the following *internals* parsers only: dhcp, ftp, mqtt, sip and ssh
- If this PR is OK for you, more PRs will follow for other internal parsers, and for parsers relying on external crate.
- Note: even though nom 7 advertises the minimum rust version is 1.48, this PR does not require a change in the minimum rust version

### CODE REVIEW

**Review by jasonish — COMMENTED** [review]
Great work, I really like these new functional parsers over the macros. I don't really have any comments on the quality of the work, but more about style:
- How much work to rustfmt, then apply your changes?  I realize this could be a pain, and getting this in is more important than this style issue.
- I don't think files like `mqtt.rs` where only an import changed should have been reformatted.  Its OK in files where the parsers make up most of the code.

**Comment by chifflier** [comment]
> Great work, I really like these new functional parsers over the macros. I don't really have any comments on the quality of the work, but more about style:
> 
>     * How much work to rustfmt, then apply your changes?  I realize this could be a pain, and getting this in is more important than this style issue.

I understand, this may help clarifying the upgrade changes from the reformatting. Works for me, I'll rebase/rewrite patches after rustfmt

> 
>     * I don't think files like `mqtt.rs` where only an import changed should have been reformatted.  Its OK in files where the parsers make up most of the code.

Agreed, this will be fixed if reformatting first.

File: rust/src/dhcp/parser.rs [hunk 1]
```diff
@@ -81,120 +83,124 @@ pub struct DHCPOption {
...
+        i,
+        DHCPOption {
+            code: code,
+            data: None,
+            option: DHCPOptionWrapper::ClientId(DHCPOptClientId {
+                htype: 1,
+                data: data.to_vec(),
+            }),
+        },
+    ))
```
[the remark this answers is not in the document]
Comment (chifflier) [reply rust/src/dhcp/parser.rs:139]: Yeah, I tried to be minimal in the proposed changes, but I also find this not very elegant.
Personally, I prefer to  split the two lines:
```rust
let res = DHCPOption {
            code,
            data: None,
            option: DHCPOptionWrapper::ClientId(DHCPOptClientId {
                htype: 1,
                data: data.to_vec(),
            }),
        };
Ok((i, res))
```

So the return line is a one-liner. Would this be OK for you?

[the remark this answers is not in the document]
Comment (jasonish) [reply rust/src/dhcp/parser.rs:139]: Yeah, might be a reasonable work-around (for rustfmt) for now.

**Review by jasonish — CHANGES_REQUESTED** [review]
Just setting changes requested flag. Comments already made.

**Comment by chifflier** [comment]
Hi,
@jasonish @victorjulien : is there an ETA for the rustfmt changes ? I understand this can have many side effects (like breaking currently pending PRs).
If you want, I can prepare a PR.

**Comment by jasonish** [comment]
No ETA.  Breaking existing PRs is a concern but I don't think any time is better than any others in this respect.  I could do a script that lists all files affected in open PRs and we could format subsystems at time that don't have any major work happening in them?

**Comment by chifflier** [comment]
It seems a good plan: we can commit rustfmt changes in directories/files that will not conflict at the moment, and commit the rest later.

**Comment by chifflier** [comment]
Replaced by #6572

**Pull request closed without merging**